File: backend-legacy-python/app/vpn/protocols/xdns.py
"""XDNS Tunnel — emergency fallback: proxy traffic over DNS queries.

When TCP and UDP are both blocked, DNS (port 53) usually stays open.
Uses xray Finalmask XDNS to tunnel data inside DNS packets.

WARNING: This is a LAST RESORT protocol — very slow, small MTU (~512 bytes).
Only enable when all other transports are blocked.
"""

from app.config import get_settings

from .base import ClientLink, Protocol, ServerConfig, UserCredentials, XrayInbound


class Xdns(Protocol):
    def __init__(self) -> None:
        s = get_settings()
        self._domain = s.xdns_domain
        self._enabled = s.xdns_enabled

    @property
    def name(self) -> str:
        return "xdns"

    @property
    def display_name(self) -> str:
        return "DNS Tunnel (Emergency)"

    @property
    def enabled(self) -> bool:
        return self._enabled and bool(self._domain)

    def xray_inbounds(
        self, users: list[UserCredentials], short_ids: list[str]
    ) -> list[XrayInbound]:
        if not self.enabled:
            return []
        return [
            XrayInbound(
                tag="xdns-in",
                port=53,
                protocol="dokodemo-door",
                settings={
                    "network": "udp",
                    "followRedirect": False,
                },
                stream_settings={
                    "finalmask": {"type": "xdns"},
                },
                sniffing={"enabled": True, "destOverride": ["http", "tls"]},
            )
        ]

    def xray_outbounds(self) -> list[dict]:
        if not self.enabled:
            return []
        return [
            {
                "protocol": "freedom",
                "tag": "xdns-out",
                "streamSettings": {
                    "finalmask": {"type": "xdns"},
                },
            }
        ]

    def xray_routing_rules(self) -> list[dict]:
        if not self.enabled:
            return []
        return [
            {
                "type": "field",
                "inboundTag": ["xdns-in"],
                "outboundTag": "xdns-out",
            }
        ]

    def client_links(
        self, user: UserCredentials, servers: list[ServerConfig]
    ) -> list[ClientLink]:
        """XDNS tunnel links — xray-native only, not sing-box compatible."""
        if not self.enabled:
            return []
        links: list[ClientLink] = []
        for srv in servers:
            remark = f"{srv.flag} {srv.name} DNS-Tunnel"
            # xdns:// is an xray-native URI scheme for Finalmask XDNS
            uri = (
                f"xdns://{user.uuid}@{self._domain}:53"
                f"?domain={self._domain}"
                f"&security=none"
                f"#{remark}"
            )
            links.append(
                ClientLink(
                    uri=uri,
                    protocol="xdns",
                    transport="dns",
                    server_key=srv.key,
                    remark=remark,
                )
            )
        return links

    def singbox_outbound(
        self, tag: str, server: ServerConfig, user: UserCredentials, **opts
    ) -> dict | None:
        # XDNS is NOT supported by sing-box — clients must use xray-based config
        return None

File: backend/crates/chameleon-vpn/src/protocols/xdns.rs
//! XDNS Tunnel — emergency fallback: proxy traffic over DNS queries.

use serde_json::json;
use chameleon_config::Settings;
use super::types::*;

pub struct Xdns {
    domain: String,
    is_enabled: bool,
}

impl Xdns {
    pub fn new(s: &Settings) -> Self {
        Self { domain: s.xdns_domain.clone(), is_enabled: s.xdns_enabled }
    }
}

impl Protocol for Xdns {
    fn name(&self) -> &str { "xdns" }
    fn display_name(&self) -> &str { "DNS Tunnel (Emergency)" }
    fn enabled(&self) -> bool { self.is_enabled && !self.domain.is_empty() }

    fn xray_inbounds(&self, _: &[UserCredentials], _: &[String]) -> Vec<XrayInbound> {
        if !self.enabled() { return vec![]; }
        vec![XrayInbound {
            tag: "xdns-in".into(), port: 53, protocol: "dokodemo-door".into(),
            settings: json!({"network": "udp", "followRedirect": false}),
            stream_settings: json!({"finalmask": {"type": "xdns"}}),
            sniffing: json!({"enabled": true, "destOverride": ["http", "tls"]}),
            listen: "0.0.0.0".into(),
        }]
    }

    fn xray_outbounds(&self) -> Vec<serde_json::Value> {
        if !self.enabled() { return vec![]; }
        vec![json!({"protocol": "freedom", "tag": "xdns-out", "streamSettings": {"finalmask": {"type": "xdns"}}})]
    }

    fn xray_routing_rules(&self) -> Vec<serde_json::Value> {
        if !self.enabled() { return vec![]; }
        vec![json!({"type": "field", "inboundTag": ["xdns-in"], "outboundTag": "xdns-out"})]
    }

    fn client_links(&self, user: &UserCredentials, servers: &[ServerConfig]) -> Vec<ClientLink> {
        if !self.enabled() { return vec![]; }
        servers.iter().map(|srv| {
            let remark = srv.remark("DNS-Tunnel");
            let uri = format!("xdns://{}@{}:53?domain={}&security=none#{}", user.uuid, self.domain, self.domain, remark);
            ClientLink { uri, protocol: "xdns".into(), transport: "dns".into(), server_key: srv.key.clone(), remark, is_relay: false }
        }).collect()
    }

    fn singbox_outbound(&self, _: &str, _: &ServerConfig, _: &UserCredentials, _: &OutboundOpts) -> Option<serde_json::Value> {
        None // Not supported by sing-box
    }
}
